src/include/distributed/metadata_cache.h

```diff
@@ -38,11 +38,12 @@ typedef struct
 
 	/* pg_dist_shard metadata (variable-length ShardInterval array) for this table */
 	int shardIntervalArrayLength;
-	ShardInterval *shardIntervalArray;
+	ShardInterval **sortedShardIntervalArray;
```

Comment:
> > @@ -38,11 +38,12 @@ typedef struct
> >  
> >  	/* pg_dist_shard metadata (variable-length ShardInterval array) for this table */
> >  	int shardIntervalArrayLength;
> > -	ShardInterval *shardIntervalArray;
> > +	ShardInterval **sortedShardIntervalArray;
> 
> In the code, whenever we needed shard interval arrays, we use array of
> shard interval pointers instead of array of shard intervals.

Well, I tend to agree that that's a mistake. There's, based on a quick
look, no reason for doing so. And in many of these cases we access all
of them, and there's often a good number. So spatial locality does play
a role.

> So, having that in the cache makes its utilization in the code
> smoother. I think having slower binary search is an acceptable trade
> of. Is that OK for you?

The above said, I think it's mostly a minor detail. So I think you can
decide where to go.